### PR TITLE
(wip) promise and callback for transfer approved account

### PR DESCRIPTION
**Pull request by ioanSL**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -20,9 +20,14 @@ overflow-checks = true
...
-borsh = "0.6.1"
-near-sdk = "0.9.2"
+borsh = "0.8.1"
+near-sdk = "3.1.0"
+near-contract-standards = "3.2"
 wee_alloc = { version = "0.4.5", default-features = false, features = [] }
 
+[dev-dependencies]
+near-sdk-sim = "3.2.0"
+mock-nft = { path = "mock-nft" }
```
Comment (idirall22) [inline Cargo.toml]: I get `failed to load manifest for dependency "mock-nft"
`
Please include this contract

File: src/lib.rs [hunk 2]
```diff
@@ -44,93 +66,118 @@ fn is_promise_success() -> bool {
...
+                pk,
+                ACCESS_KEY_ALLOWANCE,
+                env::current_account_id(),
+                // add_access_key allows given pk to call functions claim or create_account_and_claim
+                b"claim,create_account_and_claim".to_vec(),
+            )
+        } else {
+            panic!("Account already registered");
+        }
     }
```
Comment (idirall22) [inline src/lib.rs]: What do you think if we use an assert?
```suggestion
        assert!(!new_account, "Account already registered");
        Promise::new(env::current_account_id()).add_access_key(
              pk,
              ACCESS_KEY_ALLOWANCE,
              env::current_account_id(),
              // add_access_key allows given pk to call functions claim or create_account_and_claim
              b"claim,create_account_and_claim".to_vec(),
          )
```

File: src/lib.rs [hunk 3]
```diff
@@ -44,93 +66,118 @@ fn is_promise_success() -> bool {
...
-            .add_full_access_key(new_public_key.into())
-            .transfer(amount)
-            .then(ext_self::on_account_created_and_claimed(
-                amount.into(),
-                &env::current_account_id(),
-                NO_DEPOSIT,
-                ON_CREATE_ACCOUNT_CALLBACK_GAS,
-            ))
+        // Check if pk is in accounts lookupset
+        if self.accounts.contains(&env::signer_account_pk()) {
```
Comment (idirall22) [inline src/lib.rs]: here too, I think we should use assert, to keep the code in one scoop.

File: src/lib.rs [hunk 4]
```diff
@@ -152,43 +199,44 @@ impl LinkDrop {
...
         );
         let creation_succeeded = is_promise_success();
         if creation_succeeded {
-            Promise::new(env::current_account_id()).delete_key(env::signer_account_pk());
-        } else {
-            // In case of failure, put the amount back.
-            self.accounts
-                .insert(&env::signer_account_pk(), &amount.into());
+            //removing key access to pk
+            self.nft_contract_id = String::from("");
```
Comment (idirall22) [inline src/lib.rs]: Why the `nft_contract_id` is overwritten?

**Review by idirall22 — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 1 above)
Comment (ioanSL) [reply Cargo.toml]: I think we should remove the simulation tests. I didn't manage to reproduce the same scenarios as if I was testing on the testnet. Mai issue was that linking account_id an pk is not supported by near-sdk-sim. Maybe there is a solution for this but I was not able to find it.

File: src/lib.rs (hunk 4 above)
Comment (ioanSL) [reply src/lib.rs]: Fixed: [Commit](https://github.com/Shard-Labs/near-linkdrop/pull/1/commits/ba20ac597220379d4cba8c9ee1401a20e1ab6fda)

File: src/lib.rs (hunk 2 above)
Comment (ioanSL) [reply src/lib.rs]: Fixed: [Commit](https://github.com/Shard-Labs/near-linkdrop/pull/1/commits/ba20ac597220379d4cba8c9ee1401a20e1ab6fda)

File: src/lib.rs (hunk 3 above)
Comment (ioanSL) [reply src/lib.rs]: Fixed: [commit](https://github.com/Shard-Labs/near-linkdrop/pull/1/commits/8a8b59f92ce5ff828ce587bacdf58c96cf4480cb)

**Review by idirall22 — APPROVED** [review]

**Pull request merged**